File: src/webassembly/assembly/word-cloud-placing.ts
import { calcMinMax, tryPlaceOnFourSides } from './placing-four-sides';
import { AxisAlignedBoundingBox, QuadTree, Vector2 } from './quadtree';

export function isZero(t: f32): bool {
  return Math.abs(t) <= f32.EPSILON;
}

export function pushAt<T>(array: Array<T>, index: i32, element: T): void {
  assert(index <= array.length, 'pushAt should not exceed the array!');
  array.length += 1;
  array.copyWithin(index + 1, index, array.length - 1);
  array[index] = element;
}

export class Range {
  constructor(public readonly start: f32, public readonly end: f32) {
    assert(start <= end, 'Cant create invalid range!');
  }

  collides(other: Range): bool {
    // no equality, touching should not be colliding
    return other.start < this.end && other.end > this.start;
  }

  toString(): string {
    return '[' + this.start.toString() + ',' + this.end.toString() + ']';
  }
}

export const NULL_RANGE = new Range(0, 0);

export class TRangeSet {
  public readonly tRanges: Array<Range> = new Array<Range>();

  constructor(tRangeSize: f32) {
    if (tRangeSize !== 0) {
      this.tRanges.push(new Range(-tRangeSize, tRangeSize));
    }
  }

  clone(): TRangeSet {
    const newRange = new TRangeSet(0);
    for (let i = 0; i < this.tRanges.length; i++) {
      newRange.tRanges.push(this.tRanges[i]);
    }
    return newRange;
  }

  isEmpty(): bool {
    return this.tRanges.length === 0;
  }

  getBestTValue(optimalT: f32): f32 {
    let distance = f32.MAX_VALUE;
    let t = optimalT;
    for (let i = 0; i < this.tRanges.length; i++) {
      const range = this.tRanges[i];
      if (optimalT > range.end) {
        const dist = f32(Math.abs(optimalT - range.end));
        if (dist < distance) {
          distance = dist;
          t = range.end;
        }
      } else if (optimalT < range.start) {
        const dist = f32(Math.abs(optimalT - range.start));
        if (dist < distance) {
          distance = dist;
          t = range.start;
        }
        break; // sorted ascending -> break if before
      } else {
        t = optimalT;
        distance = 0;
        break;
      }
    }
    return t;
  }

  /**
   * @returns the possible range, where it collides
   *
   * returns null if it always collides
   *
   * returns NULL_RANGE if there is never a collision
   */
  mergeTRange(
    staticRange: Range,
    tRange: Range,
    tProjectedOnRange: f32
  ): Range | null {
    if (this.tRanges.length < 1) {
      return null;
    }
    if (isZero(tProjectedOnRange)) {
      return staticRange.collides(tRange) ? null : NULL_RANGE;
    }
    let tmin = (tRange.start - staticRange.end) / -tProjectedOnRange;
    let tmax = (staticRange.start - tRange.end) / tProjectedOnRange;
    if (tmax < tmin) {
      const temp = tmax;
      tmax = tmin;
      tmin = temp;
    }
    if (
      tmax <= this.tRanges[0].start ||
      tmin >= this.tRanges[this.tRanges.length - 1].end
    ) {
      return NULL_RANGE;
    }
    return new Range(tmin, tmax);
  }

  splitByRange(start: f32, end: f32): void {
    for (let i = this.tRanges.length - 1; i >= 0; i--) {
      const currentRange = this.tRanges[i];
      if (start <= currentRange.start) {
        if (end >= currentRange.end) {
          this.tRanges.splice(i, 1);
        } else if (end > currentRange.start) {
          this.tRanges[i] = new Range(end, currentRange.end);
        }
        continue;
      }
      // From here: No more collisions possible, always break
      if (start >= currentRange.end) {
        break;
      }
      if (end < currentRange.end) {
        pushAt(this.tRanges, i + 1, new Range(end, currentRange.end));
      }

      this.tRanges[i] = new Range(currentRange.start, start);
      break;
    }
  }
}

class SideAccessInformation {
  constructor(
    public readonly sideNormal: Vector2,
    public readonly perpendicularNormal: Vector2,
    public readonly sideOffset: Vector2,
    public readonly distToMid: f32,
    public readonly otherDirMidDist: f32
  ) {}
}

const DEG_TO_RAD = Math.PI / 180;

class PreparedBuildInformation {
  // topLeft, rightTop, bottomRight, leftBottom
  public readonly sides: StaticArray<SideAccessInformation> =
    new StaticArray<SideAccessInformation>(4);
  public readonly normal1: Vector2;
  public readonly normal2: Vector2;

  constructor(width: f32, height: f32, rotation: f32) {
    width /= 2;
    height /= 2;
    const cos = f32(Math.cos(rotation * DEG_TO_RAD));
    const sin = f32(Math.sin(rotation * DEG_TO_RAD));
    const normalUp = new Vector2(-sin, cos);
    const normalRight = new Vector2(cos, sin);
    this.normal1 = normalUp;
    this.normal2 = normalRight;
    this.sides[0] = new SideAccessInformation(
      normalUp.clone(),
      normalRight.clone().scale(-1),
      normalUp.clone().scale(height).add(normalRight.clone().scale(-width)),
      height,
      width
    );
    this.sides[1] = new SideAccessInformation(
      normalRight.clone(),
      normalUp.clone(),
      normalRight.clone().scale(width).add(normalUp.clone().scale(height)),
      width,
      height
    );
    this.sides[2] = new SideAccessInformation(
      normalUp.clone().scale(-1),
      normalRight.clone(),
      normalUp.clone().scale(-height).add(normalRight.clone().scale(width)),
      height,
      width
    );
    this.sides[3] = new SideAccessInformation(
      normalRight.clone().scale(-1),
      normalUp.clone().scale(-1),
      normalRight.clone().scale(-width).add(normalUp.clone().scale(-height)),
      width,
      height
    );
  }
}

class PlacedTopic {
  public readonly points: StaticArray<Vector2> = new StaticArray<Vector2>(4);
  public readonly normal1Range: Range;
  public readonly normal2Range: Range;
  public readonly xRange: Range;
  public readonly yRange: Range;

  constructor(
    public readonly mid: Vector2,
    public readonly normal1: Vector2,
    public readonly normal2: Vector2,
    public readonly normal1Size: f32,
    public readonly normal2Size: f32,
    public readonly distance: f32
  ) {
    this.points[0] = normal1
      .clone()
      .scale(normal1Size)
      .add(normal2.clone().scale(-normal2Size))
      .add(mid);
    this.points[1] = normal1
      .clone()
      .scale(normal1Size)
      .add(normal2.clone().scale(normal2Size))
      .add(mid);
    this.points[2] = normal1
      .clone()
      .scale(-normal1Size)
      .add(normal2.clone().scale(normal2Size))
      .add(mid);
    this.points[3] = normal1
      .clone()
      .scale(-normal1Size)
      .add(normal2.clone().scale(-normal2Size))
      .add(mid);
    const n1Dot = mid.dot(normal1);
    this.normal1Range = new Range(n1Dot - normal1Size, n1Dot + normal1Size);
    const n2Dot = mid.dot(normal2);
    this.normal2Range = new Range(n2Dot - normal2Size, n2Dot + normal2Size);
    this.xRange = calcMinMax(new Vector2(1, 0), this.points);
    this.yRange = calcMinMax(new Vector2(0, 1), this.points);
  }
}

export class PositionInfo {
  constructor(
    public readonly position: Vector2,
    public readonly distanceSquared: f32,
    public readonly normal1: Vector2,
    public readonly normal2: Vector2,
    public readonly normal1Size: f32,
    public readonly normal2Size: f32
  ) {}

  toPlacedPos(): PlacedTopic {
    return new PlacedTopic(
      this.position,
      this.normal1,
      this.normal2,
      this.normal1Size,
      this.normal2Size,
      f32(Math.sqrt(this.distanceSquared))
    );
  }
}

export class WordCloudTopic {
  public readonly buildInfo: PreparedBuildInformation;
  public position: PlacedTopic | null = null;
  constructor(
    public readonly width: f32,
    public readonly height: f32,
    public readonly rotation: f32
  ) {
    this.buildInfo = new PreparedBuildInformation(width, height, rotation);
  }

  collideSAT(box: AxisAlignedBoundingBox<f32>): bool {
    const pos = this.position!;
    if (pos.yRange.start >= box.y + box.height || pos.yRange.end <= box.y)
      return false;
    if (pos.xRange.start >= box.x + box.width || pos.xRange.end <= box.x)
      return false;
    if (!pos.normal1Range.collides(calcMinMax(pos.normal1, box.points)))
      return false;
    if (!pos.normal2Range.collides(calcMinMax(pos.normal2, box.points)))
      return false;
    return true;
  }

  /**
   * Only callable when current topic has a position
   * @param otherTopic topic to place around this topic
   * @returns minimal distance with position
   */
  tryPlace(
    quadTree: QuadTree<f32, WordCloudTopic>,
    otherTopic: WordCloudTopic,
    aspectRatio: f32
  ): PositionInfo | null {
    assert(this.position !== null, 'Current topic must have a position!');
    let diffAngle = (this.rotation - otherTopic.rotation) % 90;
    if (isZero(diffAngle) || isZero(diffAngle - 90)) {
      // perpendicular or colinear
      return tryPlaceOnFourSides(this, otherTopic, quadTree, aspectRatio);
    }
    assert(false, 'For testing purpose, should not happen.');
    return null;
  }
}

export function findBestPlace(
  index: i32,
  newTopic: WordCloudTopic,
  elements: StaticArray<WordCloudTopic>,
  tree: QuadTree<f32, WordCloudTopic>,
  aspectRatio: f32
): void {
  if (index === 0) {
    const side = newTopic.buildInfo.sides[0];
    newTopic.position = new PlacedTopic(
      new Vector2(0, 0),
      side.sideNormal,
      side.perpendicularNormal,
      side.distToMid,
      side.otherDirMidDist,
      0
    );
    tree.addElement(newTopic);
    return;
  }
  let posInfo: PositionInfo | null = null;
  for (let i = 0; i < index; i++) {
    const nearestPos = elements[i].tryPlace(tree, newTopic, aspectRatio);
    if (nearestPos === null) continue;
    if (posInfo === null) {
      posInfo = nearestPos;
      continue;
    } else if (posInfo.distanceSquared > nearestPos.distanceSquared) {
      posInfo = nearestPos;
    }
  }
  if (posInfo !== null) {
    newTopic.position = posInfo.toPlacedPos();
    tree.addElement(newTopic);
  }
}

File: webassembly/frag-jetzt-wasm/src/wordCloudPlacing.rs
use std::{cell::RefCell, rc::Rc};

use crate::wordCloudPlacing::placingFourSides::tryPlaceOnFourSides;

use self::quadtree::{
    calcMinMax, AxisAlignedBoundingBox, Number, QuadTree, QuadTreeElement, Range, Vector2,
};

mod placingFourSides;
pub mod quadtree;

// Orderd increasing
pub struct TRangeSet<T: Number> {
    tRanges: Vec<Range<T>>,
}

pub enum MergeRangeResult<T: Number> {
    AlwaysCollides,
    NoCollision,
    CollisionRange(Range<T>),
}

impl<T: Number> MergeRangeResult<T> {
    fn never_collides(&self) -> bool {
        matches!(*self, MergeRangeResult::NoCollision)
    }
}

impl<T: Number> TRangeSet<T> {
    pub fn new(tRangeSize: T) -> Self {
        TRangeSet {
            tRanges: vec![Range::new(-tRangeSize, tRangeSize)],
        }
    }

    pub fn isEmpty(&self) -> bool {
        self.tRanges.len() < 1
    }

    pub fn getBestTValue(&self, optimalT: T) -> T {
        let mut distance = T::MaxValue();
        let mut t = optimalT;
        for range in self.tRanges.iter() {
            if optimalT > range.get_end() {
                let dist = optimalT - range.get_end();
                if dist < distance {
                    distance = dist;
                    t = range.get_end();
                }
            } else if optimalT < range.get_start() {
                let dist = range.get_start() - optimalT;
                if dist < distance {
                    t = range.get_start();
                }
                // Ordered increasing
                return t;
            } else {
                return optimalT;
            }
        }
        return t;
    }

    pub fn mergeTRange(
        &self,
        staticRange: &Range<T>,
        tRange: &Range<T>,
        tProjectedOnRange: T,
    ) -> MergeRangeResult<T> {
        if self.isEmpty() {
            return MergeRangeResult::AlwaysCollides;
        }
        if T::isZero(tProjectedOnRange) {
            return if staticRange.collides(tRange) {
                MergeRangeResult::AlwaysCollides
            } else {
                MergeRangeResult::NoCollision
            };
        }
        let mut tmin = (tRange.get_start() - staticRange.get_end()) / -tProjectedOnRange;
        let mut tmax = (staticRange.get_start() - tRange.get_end()) / tProjectedOnRange;
        if tmax < tmin {
            let temp = tmax;
            tmax = tmin;
            tmin = temp;
        }
        if tmax <= self.tRanges[0].get_start()
            || tmin >= self.tRanges[self.tRanges.len() - 1].get_end()
        {
            return MergeRangeResult::NoCollision;
        }
        MergeRangeResult::CollisionRange(Range::new(tmin, tmax))
    }

    pub fn splitByRange(&mut self, start: T, end: T) {
        let ranges = &mut self.tRanges;
        for i in (0..ranges.len()).rev() {
            let range = &mut ranges[i];
            if start <= range.get_start() {
                if end >= range.get_end() {
                    ranges.remove(i);
                } else if end > range.get_start() {
                    range.set_start(end);
                }
                continue;
            }
            // From here: No more collisions possible, always break
            if start >= range.get_end() {
                break;
            }
            let range_end = range.get_end();
            range.set_end(start);
            if end < range_end {
                ranges.insert(i + 1, Range::new(end, range_end));
            }
            break;
        }
    }
}

struct SideAccessInformation<T: Number> {
    sideNormal: Vector2<T>,
    perpendicularNormal: Vector2<T>,
    sideOffset: Vector2<T>,
    distToMid: T,
    otherDirMidDist: T,
}

impl<T: Number> SideAccessInformation<T> {
    pub fn new(
        sideNormal: Vector2<T>,
        perpendicularNormal: Vector2<T>,
        sideOffset: Vector2<T>,
        distToMid: T,
        otherDirMidDist: T,
    ) -> Self {
        SideAccessInformation {
            sideNormal,
            perpendicularNormal,
            sideOffset,
            distToMid,
            otherDirMidDist,
        }
    }
}

struct PreparedBuildInformation<T: Number> {
    sides: Vec<SideAccessInformation<T>>,
    normal1: Vector2<T>,
    normal2: Vector2<T>,
}

impl<T: Number> PreparedBuildInformation<T> {
    pub fn new(width: T, height: T, rotation: T) -> Self {
        let width = width / T::Two();
        let height = height / T::Two();
        let (sin, cos) = T::SinCos(rotation);
        let normalUp = Vector2::new(-sin, cos);
        let normalRight = Vector2::new(cos, sin);
        PreparedBuildInformation {
            sides: vec![
                SideAccessInformation::new(
                    normalUp.clone(),
                    normalRight.clone().scale(-T::One()),
                    normalUp
                        .clone()
                        .scale(height)
                        .add(&normalRight.clone().scale(-width)),
                    height,
                    width,
                ),
                SideAccessInformation::new(
                    normalRight.clone(),
                    normalUp.clone(),
                    normalRight
                        .clone()
                        .scale(width)
                        .add(&normalUp.clone().scale(height)),
                    width,
                    height,
                ),
                SideAccessInformation::new(
                    normalUp.clone().scale(-T::One()),
                    normalRight.clone(),
                    normalUp
                        .clone()
                        .scale(-height)
                        .add(&normalRight.clone().scale(width)),
                    height,
                    width,
                ),
                SideAccessInformation::new(
                    normalRight.clone().scale(-T::One()),
                    normalUp.clone().scale(-T::One()),
                    normalRight
                        .clone()
                        .scale(-width)
                        .add(&normalUp.clone().scale(-height)),
                    width,
                    height,
                ),
            ],
            normal1: normalUp,
            normal2: normalRight,
        }
    }
}

pub struct PlacedTopic<T: Number> {
    pub mid: Vector2<T>,
    normal1: Vector2<T>,
    normal2: Vector2<T>,
    points: Vec<Vector2<T>>,
    normal1Range: Range<T>,
    normal2Range: Range<T>,
    xRange: Range<T>,
    yRange: Range<T>,
}

impl<T: Number> PlacedTopic<T> {
    pub fn new(
        mid: Vector2<T>,
        normal1: Vector2<T>,
        normal2: Vector2<T>,
        normal1Size: T,
        normal2Size: T,
        distance: T,
    ) -> Self {
        let points = vec![
            normal1
                .clone()
                .scale(normal1Size)
                .add(&normal2.clone().scale(-normal2Size))
                .add(&mid)
                .clone(),
            normal1
                .clone()
                .scale(normal1Size)
                .add(&normal2.clone().scale(normal2Size))
                .add(&mid)
                .clone(),
            normal1
                .clone()
                .scale(-normal1Size)
                .add(&normal2.clone().scale(normal2Size))
                .add(&mid)
                .clone(),
            normal1
                .clone()
                .scale(-normal1Size)
                .add(&normal2.clone().scale(-normal2Size))
                .add(&mid)
                .clone(),
        ];
        let n1Dot = mid.dot(&normal1);
        let n2Dot = mid.dot(&normal2);
        PlacedTopic {
            mid,
            normal1,
            normal2,
            normal1Range: Range::new(n1Dot - normal1Size, n1Dot + normal1Size),
            normal2Range: Range::new(n2Dot - normal2Size, n2Dot + normal2Size),
            xRange: calcMinMax(&Vector2::dir_x(), &points),
            yRange: calcMinMax(&Vector2::dir_y(), &points),
            points,
        }
    }
}

pub struct PositionInfo<T: Number> {
    position: Vector2<T>,
    distanceSquared: T,
    normal1: Vector2<T>,
    normal2: Vector2<T>,
    normal1Size: T,
    normal2Size: T,
}

impl<T: Number> PositionInfo<T> {
    pub fn new(
        position: Vector2<T>,
        distanceSquared: T,
        normal1: Vector2<T>,
        normal2: Vector2<T>,
        normal1Size: T,
        normal2Size: T,
    ) -> Self {
        PositionInfo {
            position,
            distanceSquared,
            normal1,
            normal2,
            normal1Size,
            normal2Size,
        }
    }

    pub fn toPlacedPos(&self) -> PlacedTopic<T> {
        PlacedTopic::new(
            self.position.clone(),
            self.normal1.clone(),
            self.normal2.clone(),
            self.normal1Size,
            self.normal2Size,
            T::Sqrt(self.distanceSquared),
        )
    }
}

pub trait WordCloudElement<T: Number>: QuadTreeElement<T> {
    fn get_collison_info(
        &self,
    ) -> (
        &[Vector2<T>],
        T,
        Vector2<T>,
        &Range<T>,
        Vector2<T>,
        &Range<T>,
    );
}

pub struct WordCloudObstacle<T: Number> {
    aabb: AxisAlignedBoundingBox<T>,
    xRange: Range<T>,
    yRange: Range<T>,
}

impl<T: Number> WordCloudObstacle<T> {
    pub fn new(x: T, y: T, width: T, height: T) -> Self {
        WordCloudObstacle {
            aabb: AxisAlignedBoundingBox::new(x, y, width, height),
            xRange: Range::new(x, x + width),
            yRange: Range::new(y, y + height),
        }
    }
}

impl<T: Number> QuadTreeElement<T> for WordCloudObstacle<T> {
    fn collidesWithTree(&self, aabb: &AxisAlignedBoundingBox<T>) -> bool {
        self.yRange.get_start() < aabb.y + aabb.height
            && self.yRange.get_end() > aabb.y
            && self.xRange.get_start() < aabb.x + aabb.width
            && self.xRange.get_end() > aabb.x
    }
}

impl<T: Number> WordCloudElement<T> for WordCloudObstacle<T> {
    fn get_collison_info(
        &self,
    ) -> (
        &[Vector2<T>],
        T,
        Vector2<T>,
        &Range<T>,
        Vector2<T>,
        &Range<T>,
    ) {
        (
            &self.aabb.points,
            T::Zero(),
            Vector2::dir_y(),
            &self.yRange,
            Vector2::dir_x(),
            &self.xRange,
        )
    }
}

pub struct WordCloudTopic<T: Number> {
    width: T,
    height: T,
    rotation: T,
    buildInfo: PreparedBuildInformation<T>,
    pub place: Option<PlacedTopic<T>>,
}

impl<T: Number> WordCloudTopic<T> {
    pub fn new(width: T, height: T, rotation: T) -> Self {
        WordCloudTopic {
            width,
            height,
            rotation,
            buildInfo: PreparedBuildInformation::new(width, height, rotation),
            place: None,
        }
    }

    pub fn tryPlace(
        &self,
        quadTree: &QuadTree<T>,
        otherTopic: &Self,
        aspectRatio: T,
    ) -> Option<PositionInfo<T>> {
        if self.place.is_none() {
            panic!("Current topic must have a position!");
        }
        let diffAngle = (self.rotation - otherTopic.rotation) % T::Ninety();
        if T::isZero(diffAngle) || T::isZero(diffAngle - T::Ninety()) {
            // perpendicular or colinear
            return tryPlaceOnFourSides(self, otherTopic, quadTree, aspectRatio);
        }
        panic!("For testing purpose, should not happen.");
    }
}

impl<T: Number> QuadTreeElement<T> for WordCloudTopic<T> {
    fn collidesWithTree(&self, aabb: &AxisAlignedBoundingBox<T>) -> bool {
        let pos = self.place.as_ref().unwrap();
        pos.yRange.get_start() < aabb.y + aabb.height
            && pos.yRange.get_end() > aabb.y
            && pos.xRange.get_start() < aabb.x + aabb.width
            && pos.xRange.get_end() > aabb.x
            && pos
                .normal1Range
                .collides(&calcMinMax(&pos.normal1, &aabb.points))
            && pos
                .normal2Range
                .collides(&calcMinMax(&pos.normal2, &aabb.points))
    }
}

impl<T: Number> WordCloudElement<T> for WordCloudTopic<T> {
    fn get_collison_info(
        &self,
    ) -> (
        &[Vector2<T>],
        T,
        Vector2<T>,
        &Range<T>,
        Vector2<T>,
        &Range<T>,
    ) {
        let pos = self.place.as_ref().unwrap();
        (
            &pos.points,
            self.rotation,
            pos.normal1.clone(),
            &pos.normal1Range,
            pos.normal2.clone(),
            &pos.normal2Range,
        )
    }
}

pub fn findBestPlace<T: Number + 'static>(
    new_topic: Rc<RefCell<WordCloudTopic<T>>>,
    elements: &Vec<Rc<RefCell<WordCloudTopic<T>>>>,
    tree: &mut QuadTree<T>,
    aspectRatio: T,
) {
    let newTopic = Rc::clone(&new_topic);
    let mut topic = new_topic.borrow_mut();
    if elements.len() == 0 {
        let side = &topic.buildInfo.sides[0];
        topic.place = Some(PlacedTopic::new(
            Vector2::new(T::Zero(), T::Zero()),
            side.sideNormal.clone(),
            side.perpendicularNormal.clone(),
            side.distToMid,
            side.otherDirMidDist,
            T::Zero(),
        ));
        tree.addElement(newTopic);
        return;
    }
    let mut posInfo: Option<PositionInfo<T>> = None;
    for i in 0..elements.len() {
        let nearestPos = elements[i].borrow().tryPlace(tree, &topic, aspectRatio);
        if let Some(ref pos1) = nearestPos {
            if let Some(ref pos2) = posInfo {
                if pos2.distanceSquared > pos1.distanceSquared {
                    posInfo = nearestPos;
                }
            }else{
                posInfo = nearestPos;
            }
        }
    }
    if posInfo.is_some() {
        topic.place = Some(posInfo.unwrap().toPlacedPos());
        tree.addElement(newTopic);
    }
    return;
}
